File: zroky-sdk/zroky/_protect.py
# SPDX-License-Identifier: FSL-1.1-MIT
# Copyright 2026 Zroky AI

"""High-level protected action facade.

``protect`` is the public control-plane primitive: agents submit a real-world
action intent, Zroky applies policy/approval, and the backend-owned runner and
verifier produce proof.
"""
from __future__ import annotations

from datetime import datetime
from typing import Any, Mapping

from zroky._errors import ZrokyVerifiedActionError
from zroky._verified_action import await_action_proof, verified_action


def _default_contract_version(action: str) -> str:
    return f"{action}/1.0"


def protect(
    *,
    action: str,
    params: Mapping[str, Any] | None = None,
    operation_kind: str = "EXECUTE",
    contract_version: str | None = None,
    verification_profile: str | None = None,
    execution_request: Mapping[str, Any] | None = None,
    agent_id: str | None = None,
    environment: str = "production",
    principal: dict[str, Any] | None = None,
    actor_chain: list[dict[str, Any]] | None = None,
    purpose: dict[str, Any] | None = None,
    resource: dict[str, Any] | None = None,
    trace_context: dict[str, Any] | None = None,
    deadline: datetime | str | None = None,
    idempotency_key: str | None = None,
    raise_on_approval: bool = True,
    wait_for_receipt: bool = False,
    proof_timeout_seconds: float = 120.0,
    poll_interval_seconds: float = 2.0,
) -> dict[str, Any]:
    """Create a protected action intent and return its policy decision.

    This is intentionally action-first. Use it when an agent is about to touch a
    system of record: money movement, account access, production changes,
    outbound messages, or other side-effecting tools.

    Set ``wait_for_receipt=True`` when the caller should block until source-of-
    record verification and receipt generation reach a terminal state.
    """
    normalized_action = action.strip()
    if not normalized_action:
        raise ZrokyVerifiedActionError("[ZROKY] protect requires a non-empty action.")

    normalized_operation_kind = operation_kind.strip().upper()
    if not normalized_operation_kind:
        raise ZrokyVerifiedActionError("[ZROKY] protect requires a non-empty operation_kind.")

    decision = verified_action(
        agent_id=agent_id,
        contract_version=contract_version or _default_contract_version(normalized_action),
        action_type=normalized_action,
        operation_kind=normalized_operation_kind,
        environment=environment,
        principal=principal,
        actor_chain=actor_chain,
        purpose=purpose,
        resource=resource,
        parameters=dict(params or {}),
        execution_request=execution_request,
        verification_profile=verification_profile,
        deadline=deadline,
        trace_context=trace_context,
        idempotency_key=idempotency_key,
        raise_on_approval=raise_on_approval,
    )

    if not wait_for_receipt:
        return decision

    action_id = str(decision.get("action_id") or "")
    if not action_id:
        raise ZrokyVerifiedActionError("[ZROKY] protect could not wait for receipt without action_id.")

    proof = await_action_proof(
        action_id,
        timeout_seconds=proof_timeout_seconds,
        poll_interval_seconds=poll_interval_seconds,
    )
    return {
        "action_id": action_id,
        "decision": decision,
        "proof": proof,
        "receipt": proof.get("receipt"),
        "proof_status": proof.get("proof_status"),
        "receipt_status": proof.get("receipt_status"),
        "signature_valid": proof.get("signature_valid"),
        "evidence_id": proof.get("evidence_id"),
    }

File: zroky-sdk-js/src/protect.ts
// SPDX-License-Identifier: FSL-1.1-MIT
// Copyright 2026 Zroky AI

import type { ZrokyConfig } from "./types";
import {
  awaitActionProof,
  verifiedAction,
  ZrokyVerifiedActionError,
  type ActionProofResult,
  type VerifiedActionDecision,
} from "./verified-action";

type JsonObject = Record<string, unknown>;

export interface ProtectOptions {
  action: string;
  params?: JsonObject;
  operationKind?: string;
  contractVersion?: string;
  verificationProfile?: string;
  executionRequest?: JsonObject;
  agentId?: string;
  environment?: string;
  principal?: JsonObject;
  actorChain?: JsonObject[];
  purpose?: JsonObject;
  resource?: JsonObject;
  traceContext?: JsonObject;
  deadline?: string | Date;
  idempotencyKey?: string;
  raiseOnApproval?: boolean;
  waitForReceipt?: boolean;
  proofTimeoutMs?: number;
  pollIntervalMs?: number;
}

export interface ProtectResult extends JsonObject {
  actionId: string;
  decision: VerifiedActionDecision;
  proof: ActionProofResult;
  receipt: JsonObject | null;
  proofStatus: string;
  receiptStatus: string;
  signatureValid: boolean | null;
  evidenceId: string | null;
}

function defaultContractVersion(action: string): string {
  return `${action}/1.0`;
}

export function protect(
  options: ProtectOptions & { waitForReceipt: true },
  config?: ZrokyConfig,
): Promise<ProtectResult>;
export function protect(
  options: ProtectOptions & { waitForReceipt?: false | undefined },
  config?: ZrokyConfig,
): Promise<VerifiedActionDecision>;
export async function protect(
  options: ProtectOptions,
  config: ZrokyConfig = {},
): Promise<VerifiedActionDecision | ProtectResult> {
  const action = options.action.trim();
  if (!action) {
    throw new ZrokyVerifiedActionError("[ZROKY] protect requires a non-empty action.");
  }

  const operationKind = (options.operationKind ?? "EXECUTE").trim().toUpperCase();
  if (!operationKind) {
    throw new ZrokyVerifiedActionError("[ZROKY] protect requires a non-empty operationKind.");
  }

  const decision = await verifiedAction(
    {
      agentId: options.agentId,
      contractVersion: options.contractVersion ?? defaultContractVersion(action),
      actionType: action,
      operationKind,
      environment: options.environment ?? "production",
      principal: options.principal,
      actorChain: options.actorChain,
      purpose: options.purpose,
      resource: options.resource,
      parameters: options.params,
      executionRequest: options.executionRequest,
      verificationProfile: options.verificationProfile,
      deadline: options.deadline,
      traceContext: options.traceContext,
      idempotencyKey: options.idempotencyKey,
      raiseOnApproval: options.raiseOnApproval,
    },
    config,
  );

  if (!options.waitForReceipt) {
    return decision;
  }

  const actionId = typeof decision.action_id === "string" ? decision.action_id : "";
  if (!actionId) {
    throw new ZrokyVerifiedActionError("[ZROKY] protect could not wait for receipt without action_id.", {
      decision,
    });
  }

  const proof = await awaitActionProof(
    actionId,
    {
      timeoutMs: options.proofTimeoutMs,
      pollIntervalMs: options.pollIntervalMs,
    },
    config,
  );

  return {
    actionId,
    decision,
    proof,
    receipt: proof.receipt,
    proofStatus: proof.proofStatus,
    receiptStatus: proof.receiptStatus,
    signatureValid: proof.signatureValid,
    evidenceId: proof.evidenceId,
  };
}
